### PR TITLE
rewrote the timer module from scratch

### PR DESCRIPTION
**Pull request by md5crypt**
#### new timer module
* rewritten from scratch
* 6 new api calls
* old api still works
* better rtc precision (arithmetic errors do not accumulate)
* software watchdog
* clean code

#### api doc:
```
tmr.wdclr() -- not changed
tmr.now()   -- not changed
tmr.time()  -- not changed
tmr.delay() -- not changed
tmr.alarm() -- not changed
tmr.stop()  -- changed, see below. use tmr.unregister for old functionality

tmr.register(id, interval, mode, function)
	bind function with timer and set the interval in ms
	the mode can be:
		tmr.ALARM_SINGLE for a single run alarm
		tmr.ALARM_SEMI for a multiple single run alarm
		tmr.ALARM_AUTO for a repating alarm
	tmr.register does NOT start the timer
	tmr.alarm is a tmr.register & tmr.start macro
tmr.unregister(id)
	stop alarm, unbind function and clean up memory
	not needed for ALARM_SINGLE, as it unregisters itself
tmr.start(id)
	ret: bool
	start a alarm, returns true on success
tmr.stop(id)
	ret: bool
	stops a alarm, returns true on success
	this call dose not free any memory, to do so use tmr.unregister
	stopped alarms can be started with start
tmr.interval(id, interval)
	set alarm interval, running alarm will be restarted
tmr.state(id)
	ret: (bool, int) or nil
	returns alarm status (true=started/false=stopped) and mode
	nil if timer is unregistered
tmr.softwd(int)
	set a negative value to stop the timer
	any other value starts the timer, when the
	countdown reaches zero, the device restarts
	the timer units are seconds
```

and why, why, why are compilation warnings off? I had to #pragma them at the top of the file

### CODE REVIEW

File: app/modules/tmr.c [hunk 1]
```diff
@@ -1,232 +1,358 @@
-// Module for interfacing with timer
+/*guys, srsly, turn on warnings in the makefile*/
```
Comment (creationix) [inline app/modules/tmr.c]: +1

**Comment by TerryE** [comment]
This issue of not having tighter compiler checking (e.g. `-Wall -Wextra -Wunused-parameter`) by default  and error-free makes is a excellent one and one where we the community could contribute.

However given that this @md5crypt version is a complete rewrite with potentially its own set of bugs and possible subtle changes to the the existing functionality, I would suggest that we should defer merging it into dev until at least one other developer has done a complete code review and validated testing.

**Comment by md5crypt** [comment]
well, I was shocked that it just got merged without any discussion whatsoever ^^
As I already wrote in https://github.com/nodemcu/nodemcu-firmware/issues/517 tmr.time() is broken, but that's the esp's api fault. I think a per-chip long-run RTC clock factor calculation is needed.

**Comment by marcelstoer** [comment]
> I was shocked that it just got merged without any discussion whatsoever

+1

It made me create #516.

**Pull request merged**